### DISCUSSION TITLE
Accessing Pocketbase SQLite connection or DAO from parallel service

### DISCUSSION THREAD

**Discussion by deyvidm** [pocketbase/pocketbase#2382] [Q&A] [answered]
My application (built with Pocketbase) needs an asynchronous event processor service (like [Asynq](https://github.com/hibiken/asynq)). This event processor needs read (and likely write) access to the SQLite db embedded by Pocketbase. 

After looking at other discussion posts, I understand I can safely open the db file from other services for reading, but not writing. 
Are there any ways to give Asynq access to the database without race condition risks if Pocketbase and Asynq run on the same binary? i.e. if the same `main.go` file starts both Pocketbase and Asynq? 

Ideally I would like to continue using Pocketbase, I think it's a great project. 
If I am not able to find a way, I will be forced to rewrite it all with Gin/Gorm and it would be a lot of boilerplate work.

**Comment by ganigeorgiev** [comment]
You can safely access the `app.Dao()` concurrently (both for read and writes). Writes and transactions will queue up automatically since we keep only 1 active write/transaction connection at a time.

Could you elaborate or provide a code sample what you are trying to do?

**Comment by deyvidm** [comment]
Thank you for the fast response, I will try to elaborate. I think I've already spotted holes in my design... 

Here are a couple code samples. I trimmed logging and error checking calls to make the code easier to skim.

```
// main.go

func main() {

	// setup pocketbase (website backend API + embedded DB access)
	app := pocketbase.New()

	// setup asynq (custom event scheduling and response-handling server)
	srv := asynq.NewServer(...)
	mux := asynq.NewServeMux()
	mux.HandleFunc(task.TypeReminderEmail, task.HandleReminderEmailTask)

	// wait groups to run asynq + pocketbase concurrently
	wg := new(sync.WaitGroup)
	wg.Add(2)

	// start asynq
	go func() {
		srv.Run(mux)
		wg.Done()
	}()

	// start pocketbase
	go func() {
		app.Start()
		wg.Done()
	}()

	wg.Wait()

}
```

The following file holds my Asynq handlers.
```
// task.go

func HandleReminderEmailTask(ctx context.Context, t *asynq.Task) error {
	var p emailTaskPayload
	json.Unmarshal(t.Payload(), &p)
	...
	return nil
}
```


I planned to emit a task from a Pocketbase event hook which would then be handled by this Asynq Event Handler.
Such a task would be used to schedule, for example, an sms notification to be sent a week from today.
For this, Asynq needs access to the DB so that it can pull contact and message details. 

However, I am not able to inject the DAO here through function arguments, because the function signature is strictly expected by Asynq. Unless I can somehow inject the DAO through context. 

From here, my only solution in mind is to emit self-contained events to Asynq, which holds all the details that Asynq would need to operate, or to decouple Pocketbase and Asynq entirely. But with decoupling, I lose all external write access to Pocketbase's DB. 

Thank you for your time, and your help. I think this discussion is revealing some bad design on my part, and that perhaps I need to spend more time planning the structure of my project.

**Reply by ganigeorgiev** [reply, marked as answer]
> However, I am not able to inject the DAO here through function arguments, because the function signature is strictly expected by Asynq.

You can create your own wrapper function that returns another function with the expected Asynq declaration, something like:

```go
// https://pkg.go.dev/github.com/hibiken/asynq#HandlerFunc
func HandleReminderEmailTask(app core.App) asynq.HandlerFunc {
    return func (ctx context.Context, t *asynq.Task) error {
        // you now have acces to the app.*
        // ...
    } 
}

mux.HandleFunc(task.TypeReminderEmail, task.HandleReminderEmailTask(app))
```

_Side-note_: Keep in mind that `app.Dao()` is accessible only after the app has been "bootstrapped" (aka. after db open and initial settings load). To ensure that Asynq works only with initialized app instance, you can fire the Asynq goroutine inside a [`app.OnAfterBootstrap()` hook](https://pocketbase.io/docs/event-hooks/#app-hooks).

**Reply by deyvidm** [reply]
How brilliant! Thank you, Gani, for your help and your time.
